File: lee/src/main/java/org/example/lee/题目/回溯/c_N皇后.java
package org.example.lee.题目.回溯;

import java.util.ArrayList;
import java.util.List;

public class c_N皇后 {

    /**
     * 按照国际象棋的规则，皇后可以攻击与之处在同一行或同一列或同一斜线上的棋子。
     * n 皇后问题 研究的是如何将 n 个皇后放置在 n×n 的棋盘上，并且使皇后彼此之间不能相互攻击。
     * 给你一个整数 n ，返回所有不同的 n 皇后问题 的解决方案。
     * 每一种解法包含一个不同的 n 皇后问题 的棋子放置方案，该方案中 'Q' 和 '.' 分别代表了皇后和空位。
     * solve nqueens
     * 回溯第一题 不是哥们 这真牛阿
     * @param n n
     * @return {@link List}<{@link List}<{@link String}>>
     */
    public static List<List<String>> solveNQueens(int n) {
        char[][] arrs = new char[n][n];
        List<List<String>> list = new ArrayList<>();
        for (char[] arr:
             arrs) {
            for (int i = 0; i < n; i++) {
                arr[i]='.';
            }
        }
        solutions(list,arrs,0,n);
        return list;
    }

    public static void solutions(List<List<String>> list, char[][] arr, int row, int n) {
        if (row == n) {
            ArrayList<String> strings = new ArrayList<>();
            for (char[] c : arr) {
                strings.add(new String(c));
            }
            list.add(strings);
            return;
        }
        for (int i = 0; i < n; i++) {
            if (isValid(arr,i,row,n)) {
                arr[row][i]='Q';
                solutions(list,arr,row+1,n);
                arr[row][i]='.';
            }
        }
    }

    public static boolean isValid(char[][] arr,int columns,int row,int n) {
        for (int i = 0; i < row; i++) {
            if(arr[i][columns] == 'Q'){ return false;}
        }
        for (int a = row-1,b = columns-1; a >=0 && b>=0 ; a--,b--) {
             if(arr[a][b] == 'Q'){return false;}
        }
        for (int a = row-1,b=columns+1; a >=0&& b<n ;a--,b++) {
            if(arr[a][b] == 'Q'){return false;}
        }
        return true;
    }
}

File: lee/src/main/java/org/example/lee/题目/回溯/c_N皇后.go
package 回溯

func solveNQueens(n int) [][]string {
	var awe [][]string
	arrr := make([][]byte, n)
	for i := 0; i < n; i++ {
		arrr[i] = make([]byte, n) // 为每个内层切片分配内存
		for j := 0; j < n; j++ {
			arrr[i][j] = '.'
		}
	}
	solveTool(n, 0, &awe, []string{}, arrr)
	return awe
}

func solveTool(n, row int, awe *[][]string, arr []string, arrr [][]byte) {
	if n == row {
		temp := make([]string, len(arr))
		copy(temp, arr)
		*awe = append(*awe, temp)
		return
	}
	for i := 0; i < n; i++ {
		if solveHelper(arrr, row, i) {
			arrr[row][i] = 'Q'
			arr = append(arr, string(arrr[row]))
			solveTool(n, row+1, awe, arr, arrr)
			arrr[row][i] = '.'
			arr = arr[:len(arr)-1]
		}
	}
}
func solveHelper(arrr [][]byte, x, y int) bool {
	for i := 0; i < x; i++ {
		if arrr[i][y] == 'Q' {
			return false
		}
	}
	for i, j := x-1, y-1; i >= 0 && j >= 0; i, j = i-1, j-1 {
		if arrr[i][j] == 'Q' {
			return false
		}
	}
	for i, j := x-1, y+1; i >= 0 && j < len(arrr[0]); i, j = i-1, j+1 {
		if arrr[i][j] == 'Q' {
			return false
		}
	}
	return true
}
